api/src/feeds/impl/search_api_impl.py

```diff
@@ -0,0 +1,82 @@
+from database_gen.sqlacodegen_models import t_feedsearch
+from feeds.impl.models.search_feeds200_response_results_inner_impl import SearchFeeds200ResponseResultsInnerImpl
+from feeds_gen.apis.search_api_base import BaseSearchApi
+from feeds_gen.models.search_feeds200_response import SearchFeeds200Response
+
+feed_search_columns = [column for column in t_feedsearch.columns if column.name != 'document']
+
+
+class SearchApiImpl(BaseSearchApi):
+    """
+    This class represents the implementation of the `/search` endpoints.
+    """
+
+    @staticmethod
+    def add_search_query_filters(query, search_query, data_type, feed_id, status) -> Query:
+        if feed_id:
+            query = query.where(t_feedsearch.c.feed_stable_id == feed_id.strip().lower())
+        if data_type:
+            query = query.where(t_feedsearch.c.data_type == data_type.strip().lower())
+        if status:
+            query = query.where(t_feedsearch.c.status == status.strip().lower())
+        if search_query:
+            query = query.filter(t_feedsearch.c.document.op('@@')(func.plainto_tsquery('english', f'{search_query}:*')))
+        return query
+
+    @staticmethod
+    def create_count_search_query(status: str,
+                                  feed_id: str,
+                                  data_type: str,
+                                  search_query: str) -> Query:
+        """
+        Create a search query for the database.
+        """
+        query = (select([func.count(t_feedsearch.c.feed_id)]))
+        return SearchApiImpl.add_search_query_filters(query, search_query, data_type, feed_id, status)
+
+    @staticmethod
+    def create_search_query(status: str,
+                            feed_id: str,
+                            data_type: str,
+                            search_query: str) -> Query:
+        """
+        Create a search query for the database.
+        """
+        # TODO: Add sorting and keep the rank sorting by default
+        rank_expression = func.ts_rank(t_feedsearch.c.document, func.to_tsquery('english', f'{search_query}:*')).label(
+            'rank')
+        query = (select([
+            rank_expression,
+            *feed_search_columns,
+        ]))
+        query = SearchApiImpl.add_search_query_filters(query, search_query, data_type, feed_id, status)
+        return query.order_by(rank_expression.desc())
+
+    def search_feeds(
+            self,
+            limit: int,
+            offset: int,
+            status: str,
+            feed_id: str,
+            data_type: str,
+            search_query: str,
+    ) -> SearchFeeds200Response:
+        """Search feeds using full-text search on feed, location and provider&#39;s information."""
+        query = self.create_search_query(status, feed_id, data_type, search_query)
+        feed_rows = Database().select(
+            query=query,
+            limit=limit,
+            offset=offset,
+        )
+        feed_total_count = Database().select(
```

Comment:
wouldn't it be better to have 1 database class instance and reusing it vs 2 `Database()`? Also whats the protocol to initiate a session directly in `database.by` instead of using `helpers/database.py`

Reply:
The `helpers/database.py` is a different implementation of the database connection. It belongs to the Python functions and is not part of the API source., even though it is in the same repository. In the future, we should try to unify the API with the Python functions to make the development process less confusing.